File: LogicAndTrick.WikiCodeParser/State.cs
using System;

namespace LogicAndTrick.WikiCodeParser
{
    /// <summary>
    /// Character-indexed state tracker for a string
    /// </summary>
    public class State
    {
        public string Text { get; }
        public int Length => Text.Length;
        public int Index { get; private set; }
        public bool Done => Index >= Length;

        public State(string text)
        {
            Text = text;
            Index = 0;
        }

        public string ScanTo(string find, StringComparison comparison = StringComparison.Ordinal)
        {
            var pos = Text.IndexOf(find, Index, comparison);
            if (pos < 0) pos = Length;
            var ret = Text.Substring(Index, pos - Index);
            Index = pos;
            return ret;
        }

        public void SkipWhitespace()
        {
            while (Index < Length && Char.IsWhiteSpace(Text[Index])) Index++;
        }

        public string PeekTo(string find)
        {
            var pos = Text.IndexOf(find, Index, StringComparison.Ordinal);
            if (pos < 0) return null;
            return Text.Substring(Index, pos - Index);
        }

        public void Seek(int index, bool fromStart)
        {
            Index = fromStart ? index : Index + index;
        }

        public string Peek(int count)
        {
            if (Index + count > Length) count = Length - Index;
            return Text.Substring(Index, count);
        }

        public char Next()
        {
            return Text[Index++];
        }

        public string GetToken()
        {
            if (Done || Text[Index] != '[') return null;
            var found = false;
            var tok = "";
            for (var i = Index + 1; i < Math.Min(Index + 10, Length); i++)
            {
                var c = Text[i];
                if (c == ' ' || c == '=' || c == ']')
                {
                    found = tok.Length > 0;
                    break;
                }

                tok += c;
            }

            return found ? tok.ToLower() : null;
        }
    }
}

File: src/State.ts

export class State {

    public readonly Text: string;
    public get Length(): number {
        return this.Text.length;
    }
    public Index: number;
    public get Done(): boolean {
        return this.Index >= this.Length;
    }

    public constructor(text: string) {
        this.Text = text;
        this.Index = 0;
    }

    public ScanTo(find: string, ignoreCase = false): string {
        let pos = ignoreCase
            ? this.Text.toLowerCase().indexOf(find.toLowerCase(), this.Index)
            : this.Text.indexOf(find, this.Index);
        if (pos < 0) pos = this.Length;
        const ret = this.Text.substring(this.Index, pos);
        this.Index = pos;
        return ret;
    }

    public SkipWhitespace() {
        while (this.Index < this.Length && this.Text[this.Index].trim() == '') this.Index++;
    }

    public PeekTo(find: string): string | null {
        const pos = this.Text.indexOf(find, this.Index);
        if (pos < 0) return null;
        return this.Text.substring(this.Index, pos);
    }

    public Seek(index: number, fromStart: boolean) {
        this.Index = fromStart ? index : this.Index + index;
    }

    public Peek(count: number): string {
        if (this.Index + count > this.Length) count = this.Length - this.Index;
        return this.Text.substring(this.Index, this.Index + count);
    }

    public Next(): string {
        if (this.Index >= this.Length) return '\0';
        return this.Text[this.Index++];
    }

    public GetToken(): string {
        if (this.Done || this.Text[this.Index] != '[') return null;
        let found = false;
        let tok = '';
        for (let i = this.Index + 1; i < Math.min(this.Index + 10, this.Length); i++) {
            const c = this.Text[i];
            if (c == ' ' || c == '=' || c == ']') {
                found = tok.length > 0;
                break;
            }

            tok += c;
        }

        return found ? tok.toLowerCase() : null;
    }
}
